src/phylo_flags.hpp

```diff
@@ -0,0 +1,336 @@
+  // `substitution_model_rates` key).
+  std::string key_;
+};
+
+// ** Phylo Mapkey Set
+// Contains all possible options for function.
+class PhyloMapkeySet {
+ public:
+  using MapkeyMap = std::map<std::string, PhyloMapkey>;
+
+  explicit PhyloMapkeySet(const std::string &name) : name_(name){};
+  PhyloMapkeySet(const std::string &name, const std::vector<PhyloMapkey> &mapkeys);
+
+  // Insert individual mapkey.
+  void AddMapkey(const PhyloMapkey &mapkey, const bool overwrite = false);
+  // Does mapkey already exist in set?
+  bool ContainsMapkey(const PhyloMapkey &mapkey);
+  const MapkeyMap &GetMapkeys() const { return all_mapkeys_; };
+  std::string ToString() const;
+
+ private:
+  // Name for mapkey set.
+  std::string name_;
+  // List of all possible keys.
+  MapkeyMap all_mapkeys_;
+};
+
+// ** Phylo FlagOption
+// This is the base option flag type.  Also specifies default behaviour for flags.
+class PhyloFlagOption {
+ public:
+  enum class FlagType { None, Include, Exclude, SetValue, MapKey, RunAll };
+  enum class DataType { None, Bool, Double, String };
+
+  PhyloFlagOption();
+  PhyloFlagOption(const std::string &name, const std::string &flag,
+                  const bool is_set_when_running_defaults, const FlagType flag_type,
+                  const DataType data_type);
+  // PhyloFlagOption FlagType specific constructors.
+  static PhyloFlagOption IncludeOption(const std::string &name, const std::string &flag,
+                                       const bool is_set_when_running_defaults = true);
+  static PhyloFlagOption ExcludeOption(const std::string &name, const std::string &flag,
+                                       const bool is_set_when_running_defaults = false);
+  static PhyloFlagOption SetValueOption(const std::string &name,
+                                        const std::string &flag,
+                                        const DataType data_type);
+  static PhyloFlagOption MapKey(const std::string &name, const std::string &flag);
+
+  // Add Child Flags (these flags are set when Parent is set).
+  void AddChild(const PhyloFlagOption &child);
+  void AddChild(const std::string child_flag);
+  // Output to String.
+  std::string ToString() const;
+  // Comparators
+  static int Compare(const PhyloFlagOption &flag_a, const PhyloFlagOption &flag_b);
+  // General compare.
+  bool operator==(const PhyloFlagOption &other);
+  friend bool operator==(const PhyloFlagOption &lhs, const PhyloFlagOption &rhs);
+  bool operator<(const PhyloFlagOption &other);
+  friend bool operator<(const PhyloFlagOption &lhs, const PhyloFlagOption &rhs);
+  // Compare against String
+  bool operator==(const std::string &other_name);
+  bool operator<(const std::string &other_name);
+
+  // Getters
+  std::string GetName() const { return name_; };
+  std::string GetFlag() const { return flag_; };
+  std::string operator()() const { return GetFlag(); };
+  bool GetDefaultBehavior() const { return is_set_when_running_defaults_; };
```

Comment:
`IsSetWhenRunningDefaults` seems like a more clear name, don't you think?

Reply:
Renamed.